tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11448,6 +11464,103 @@ void DiagnoseEntryAttrAllowedOnStage(clang::Sema *self,
+      }
+      case clang::attr::HLSLNodeLaunch: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeLaunch"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeIsProgramEntry: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeIsProgramEntry"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeId: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeId"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeLocalRootArgumentsTableIndex: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeLocalRootArgumentsTableIndex"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeShareInputOf: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeShareInputOf"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeDispatchGrid: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeDispatchGrid"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeMaxDispatchGrid: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeMaxDispatchGrid"
+              << "node";
+        }
+        break;
+      }
+      case clang::attr::HLSLNodeMaxRecursionDepth: {
+        switch (shaderKind) {
+        case DXIL::ShaderKind::Node:
+          break;
+        default:
+          self->Diag(pAttr->getRange().getBegin(),
+                     diag::err_hlsl_attribute_unsupported_stage)
+              << "NodeMaxRecursionDepth"
+              << "node";
+        }
+        break;
```

Comment:
```suggestion
      case clang::attr::HLSLNodeLaunch:
      case clang::attr::HLSLNodeIsProgramEntry:
      case clang::attr::HLSLNodeId:
      case clang::attr::HLSLNodeLocalRootArgumentsTableIndex:
      case clang::attr::HLSLNodeShareInputOf:
      case clang::attr::HLSLNodeDispatchGrid:
      case clang::attr::HLSLNodeMaxDispatchGrid:
      case clang::attr::HLSLNodeMaxRecursionDepth: {
        switch (shaderKind) {
        case DXIL::ShaderKind::Node:
          break;
        default:
          self->Diag(pAttr->getRange().getBegin(),
                     diag::err_hlsl_attribute_unsupported_stage)
              << pAttr->getName()
              << "node";
        }
        break;
```

We should be able to collapse this all down.